File: classic150/Solution0022.cpp
#include <vector>
#include <string>
#include <array>
#include <memory>
using namespace std;
// 数字 n 代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且 有效的 括号组合。

// 示例 1：

// 输入：n = 3
// 输出：["((()))","(()())","(())()","()(())","()()()"]
// 示例 2：

// 输入：n = 1
// 输出：["()"]

// 提示：

// 1 <= n <= 8
class Solution
{
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> result;
        backtrack(result, "", 0, 0, n);
        return result;
    }

private:
    void backtrack(vector<string> &result, string path, int left, int right, int n)
    {
        if (path.size() == n * 2)
        {
            result.push_back(path);
            return;
        }
        if (left < n)
            backtrack(result, path + "(", left + 1, right, n);
        if (right < left)
            backtrack(result, path + ")", left, right + 1, n);
    }
};

class Solution2
{
public:
    vector<string> generateParenthesis(int n)
    {
        vector<string> result;
        recursive(result, "", n);
        return result;
    }

private:
    void recursive(vector<string> result, string path, int n)
    {
        if (path.size() == n * 2)
        {
            if (isValid(path))
                result.push_back(path);
            return;
        }
        recursive(result, path + "(", n);
        recursive(result, path + ")", n);
    }
    bool isValid(string s)
    {
        int balance = 0;
        for (char c : s)
        {
            if (c == '(')
                balance++;
            else
                balance--;
            if (balance < 0)
                return false;
        }
        return balance == 0;
    }
};

class Solution3
{
private:
    array<shared_ptr<vector<string>>, 9> f;

public:
    vector<string> generateParenthesis(int n)
    {
        return *dp(n);
    }

private:
    shared_ptr<vector<string>> dp(int n)
    {
        if (f[n] != nullptr)
            return f[n];
        auto ans = make_shared<vector<string>>();
        if (n == 0)
        {
            ans->push_back("");
        }
        else
        {
            for (int c = 0; c < n; ++c)
            {
                for (auto left : *dp(c))
                {
                    for (auto right : *dp(n - 1 - c))
                    {
                        ans->push_back("(" + left + ")" + right);
                    }
                }
            }
        }
        f[n] = ans;
        return f[n];
    }
};

File: anew/Solution0022.java
package anew;

import java.util.ArrayList;
import java.util.List;

// 数字 n 代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且 有效的 括号组合。

public class Solution0022 {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(res, new StringBuilder(), 0, 0, n);
        return res;
    }

    private void backtrack(List<String> res, StringBuilder sb, int open, int close, int n) {
        if (sb.length() == 2 * n) {
            res.add(sb.toString());
            return;
        }
        if (open < n) {
            sb.append('(');
            backtrack(res, sb, open + 1, close, n);
            sb.deleteCharAt(sb.length() - 1);
        }
        if (close < open) {
            sb.append(')');
            backtrack(res, sb, open, close + 1, n);
            sb.deleteCharAt(sb.length() - 1);
        }
    }

    public List<String> generateParenthesis2(int n) {
        List<String> res = new ArrayList<>();
        recursive(res, new StringBuilder(), n);
        return res;
    }

    private void recursive(List<String> res, StringBuilder sb, int n) {
        if (sb.length() == 2 * n) {
            String str = sb.toString();
            if (valid(str)) {
                res.add(str);
            }
            return;
        }
        sb.append('(');
        recursive(res, sb, n);
        sb.deleteCharAt(sb.length() - 1);
        sb.append(')');
        recursive(res, sb, n);
        sb.deleteCharAt(sb.length() - 1);
    }

    private boolean valid(String str) {
        int balance = 0;
        for (char c : str.toCharArray()) {
            if (c == '(') {
                balance++;
            } else {
                balance--;
            }
            if (balance < 0)
                return false;
        }
        return balance == 0;
    }

    public List<String> generateParenthesis3(int n) {
        List<List<String>> cache = new ArrayList<>(n + 1);
        for (int i = 0; i <= n; i++) {
            cache.add(null);
        }
        return generate(n, cache);
    }

    private List<String> generate(int n, List<List<String>> cache) {
        if (cache.get(n) != null) {
            return cache.get(n);
        }
        List<String> res = new ArrayList<>();
        if (n == 0) {
            res.add("");
        } else {
            for (int i = 0; i < n; i++) {
                for (String left : generate(i, cache)) {
                    for (String right : generate(n - i - 1, cache)) {
                        res.add("(" + left + ")" + right);
                    }
                }
            }
        }
        cache.set(n, res);
        return res;
    }

    public static void main(String[] args) {
        Solution0022 solution = new Solution0022();
        List<String> ans = solution.generateParenthesis(3);
        System.out.println(ans);

        List<String> ans2 = solution.generateParenthesis2(3);
        System.out.println(ans2);

        List<String> ans3 = solution.generateParenthesis3(3);
        System.out.println(ans3);
    }
}
